onnx/defs/shape_inference.h

```diff
@@ -74,11 +74,27 @@ struct InferenceContext {
   virtual const SparseTensorProto* getInputSparseData(size_t index) const = 0;
 };
 
+struct DataPropagationContext {
+  virtual size_t getNumInputs() const = 0;
+  virtual const TypeProto* getInputType(size_t index) const = 0;
+  virtual size_t getNumOutputs() const = 0;
+  virtual TypeProto* getOutputType(size_t index) = 0;
+  virtual std::string getOutputName(size_t index) const = 0;
+  virtual ~DataPropagationContext() {}
+  virtual void addGeneratedShapeData(size_t index, TensorShapeProto&& tp) = 0;
+  virtual const TensorShapeProto* getGeneratedShapeData(size_t index) const = 0;
+  virtual const TensorShapeProto* getGeneratedShapeDataFromName(std::string name) const = 0;
```

Comment:
Why do we need something name-based? I think index-based should be sufficient, and is more efficient.

Reply:
Positional access is what we use in the other Contexts.

Reply:
Right now it is solely used by test (verifying the generated propagation data by  given output name). Current getPropagatedData can only get data from input index. Perhaps I should to add another get function to get from output index instead of string name.

Reply:
Yes, I think so. Further, I would suggest names like ```getInput``` or ```getInputData```.  It is also worth documenting that what is propagated are actual (partial) values, encoded using a TensorShapeProto, as opposed to shapes of inputs and outputs. E.g., for first input, we get whatever we know about the first input value (not just its shape).

Reply:
The "by-name" ability is not required for the node-level propagation-method, which doesn't even know the name of the actual-variable passed in as input or output. The tester is more like the caller, which can use the actual implementation class (which exposes more methods than the Context interface).